gnc/sub8_trajectory_generator/src/sub8_state_space.cc

```diff
@@ -18,30 +21,58 @@ void Sub8StateSpace::freeState(State* state) const {
   CompoundStateSpace::freeState(state);
 }
 
-double Sub8StateSpace::distance(const State* state1,
-                                const State* state2) const {
-  // TODO
-  // return LQR cost to go to state 2 from state 1?
+double Sub8StateSpace::distance(const State* a, const State* b) const {
+  Vector13d state_a = a->as<Sub8StateSpace::StateType>()->getState();
+  Vector13d state_b = b->as<Sub8StateSpace::StateType>()->getState();
 
-  // naive implementation is euclidean distance in position
-  Vector3d pos1;
-  Vector3d pos2;
-  state1->as<Sub8StateSpace::StateType>()->getPosition(pos1);
-  state2->as<Sub8StateSpace::StateType>()->getPosition(pos2);
+  // https://www-preview.ri.cmu.edu/pub_files/pub4/kuffner_james_2004_1/kuffner_james_2004_1.pdf
+  double lambda = state_a.segment(9, 4).dot(state_b.segment(9, 4));
 
-  return (pos2 - pos1).norm();
+  // Weighted sum of distances between position, velocity, angular velocity &
+  // Approximate "distance" metric between orientations; a.k.a angle between
+  // orientations
+  return (weights_[0] *
+          (state_b.segment(0, 3) - state_a.segment(0, 3)).norm()) +
+         (weights_[1] *
+          (state_b.segment(3, 3) - state_a.segment(3, 3)).norm()) +
+         (weights_[2] *
+          (state_b.segment(6, 3) - state_a.segment(6, 3)).norm()) +
+         (weights_[3] * acos(fabs(lambda)));
 }
 
-double Sub8StateSpace::getMaximumExtent() const {
-  // TODO
-  // PLACEHOLDER
-  return 100;
-}
+double Sub8StateSpace::getMaximumExtent() const { return INFINITY; }
+
+bool Sub8StateSpace::equalStates(const State* a, const State* b) const {
+  double error = 1e-4;
+
+  Vector13d state_a = a->as<Sub8StateSpace::StateType>()->getState();
+  Vector13d state_b = b->as<Sub8StateSpace::StateType>()->getState();
+
+  // compare position, velocity, and angular velocity
+  for (unsigned int i = 0; i < 9; ++i) {
+    if (fabs(state_a(i) - state_b(i)) > error) {
+      return false;
+    }
+  }
+
+  // compare orientation
+  // dot of two similar orientations should be 1
+  if (fabs(state_a.segment(9, 4).dot(state_b.segment(9, 4))) <
```

Comment:
This might be risky business: Since -q and q represent the same rotation for any quaternion, q, and q.dot(-q) = -1. 

This would yield false negatives in the case that q_a = -q_b.

I think your dot product idea is a good one, though - it might not be worth all of the trig to compute the difference over a lie algebra